[SPARK-36296][SQL] Refactor seventh set of 20 in QueryExecutionErrors to use error classes

dchvn opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
Refactor some exceptions in QueryExecutionErrors to use error classes.
```
missingJdbcTableNameAndQueryError
emptyOptionError
invalidJdbcTxnIsolationLevelError
cannotGetJdbcTypeError
unrecognizedSqlTypeError
unsupportedJdbcTypeError
unsupportedArrayElementTypeBasedOnBinaryError
nestedArraysUnsupportedError
cannotTranslateNonNullValueForFieldError
invalidJdbcNumPartitionsError
transactionUnsupportedByJdbcServerError
dataTypeUnsupportedYetError
unsupportedOperationForDataTypeError
inputFilterNotFullyConvertibleError
cannotReadFooterForFileError
cannotReadFooterForFileError
foundDuplicateFieldInCaseInsensitiveModeError
failedToMergeIncompatibleSchemasError
ddlUnsupportedTemporarilyError
operatingOnCanonicalizationPlanError
```


### Why are the changes needed?
There are currently ~350 exceptions in this file; so this PR only focuses on the seventh set of 20.


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Existing UT Testcase

dchvn committed:
[SPARK-36296][SQL] Refactor seventh set of 20 in QueryExecutionErrors to use error classes

AmplabJenkins commented:
Can one of the admins verify this patch?

karenfeng reviewed (commented):
Thanks for working on these @dchvn! I left some comments. Can you also add a note in the README that we should not add error classes that are not being used in Spark? I noticed while running `git grep` that some functions (eg. `operatingOnCanonicalizationPlanError`) are not being called anywhere.

phamhuyhoang97 committed:
resolve review

dchvn committed:
Merge remote-tracking branch 'upstream/master' into SPARK-36296

SparkQA commented:
**[Test build #144345 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144345/testReport)** for PR 34178 at commit [`d4133e6`](https://github.com/apache/spark/commit/d4133e6adbb42a3a3d4fdfdfc01f7ed89b72af07).

dchvn committed:
remove unused error classes

dchvn commented:
cc @karenfeng, thank you for the reviews. I updated this PR and added a note to the README file as well. I'm not sure about the note. Can you pls check if you have time? Thanks again.

SparkQA commented:
**[Test build #144348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144348/testReport)** for PR 34178 at commit [`9e00238`](https://github.com/apache/spark/commit/9e00238130f1a4c04dfcd6e3a68e07d1fd58d558).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48826/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48826/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/144348/
